src/main/java/com/google/devtools/build/lib/bazel/rules/python/python_stub_template.txt

```diff
@@ -95,8 +95,22 @@ def CreatePythonPathEntries(python_imports, module_space):
   parts = python_imports.split(':')
   return [module_space] + ['%s/%s' % (module_space, path) for path in parts]
 
-def FindModuleSpace():
+def FindModuleSpace(rel_path):
   """Finds the runfiles tree."""
+  # When the calling process used the runfiles manifest to resolve the
+  # location of this stub script, the path may be expanded. This means
+  # argv[0] may no longer point to a location inside the runfiles
+  # directory. We should therefore respect RUNFILES_DIR and
+  # RUNFILES_MANIFEST_FILE set by the caller.
+  runfiles_dir = os.environ.get('RUNFILES_DIR', None)
+  if not runfiles_dir:
+    runfiles_manifest_file = os.environ.get('RUNFILES_MANIFEST_FILE', '')
+    if (runfiles_manifest_file.endswith('.runfiles_manifest') or
+        runfiles_manifest_file.endswith('.runfiles/MANIFEST')):
+      runfiles_dir = runfiles_manifest_file[:-9]
+  if runfiles_dir and os.path.exists(os.path.join(runfiles_dir, rel_path)):
```

Comment:
This is a safety belt to ensure that we only respect RUNFILES_DIR if we know it points to a valid runfiles directory that (most likely) belongs to the Python script. If it happens to point to a disjoint runfiles directory owned by the parent, it'll end up ignoring it.

Reply:
Did you see this case occur? If so, how did it come up?

I think this logic is fine, just add a comment:
```suggestion
  # Be defensive: the runfiles dir should contain our main entry point. If
  # it doesn't, then it must not be our runfiles directory.
  if runfiles_dir and os.path.exists(os.path.join(runfiles_dir, rel_path)):
```

Reply:
Yes! I saw a bunch of shell tests in the Bazel repository itself fail if this logic wasn't present. I've added the comment you've suggested.

Reply:
Do you recall which, by chance?
I just want to look if the tests you saw fail have the same property for some tests I saw failing while I was working on this same change for our Google-version of the stub script. The thing I saw were some shell tests directly embedding a blaze-bin path directly into the tests rather than going through $0.runfiles.